avail-server/src/main/kotlin/com/avail/server/io/files/FileActions.kt

```diff
@@ -0,0 +1,347 @@
+ *   The location in the file to inserting/overwriting the data.
+ * @param end
+ *   The location in the file to stop overwriting, exclusive. All data from
+ *   this point should be preserved.
+ */
+internal class InsertRange constructor(
+	val data: ByteArray,
+	private val start: Int,
+	private val end: Int): FileAction
+{
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction =
+		file.insertRange(data, start, end, timestamp)
+
+	override val type: FileActionType = FileActionType.INSERT_RANGE
+
+	override val isTraced: Boolean = true
+}
+
+/**
+ * `NoAction` is a [FileAction] indicates no action should/could be taken.
+ * 
+ * Some `FileAction`s have an inverse action. `NoAction` is used as the inverse
+ * action to `FileAction`s that have no meaningful inverse.
+ *
+ * @author Richard Arriaga &lt;rich@availlang.org&gt;
+ */
+internal object NoAction: FileAction
+{
+	val tracedAction = TracedAction(0, NoAction, NoAction)
+
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction =
+		tracedAction
+
+	override val type: FileActionType = FileActionType.NO_ACTION
+}
+
+
+/**
+ * `UndoAction` is a [FileAction] that [executes][execute] the
+ * [inverse][TracedAction.reverseAction] for a `FileAction` that is traced as a
+ * [TracedAction].
+ *
+ * @author Richard Arriaga &lt;rich@availlang.org&gt;
+ */
+internal object UndoAction: FileAction
+{
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction
+	{
+		file.serverFileWrapper.undo()
+		return NoAction.tracedAction
+	}
+
+	override val type: FileActionType = FileActionType.UNDO
+}
+
+/**
+ * `RedoAction` is a [FileAction] that [re-executes][execute] a recently
+ * [undone][UndoAction] [traced][TracedAction] `FileAction`.
+ *
+ * @author Richard Arriaga &lt;rich@availlang.org&gt;
+ */
+internal object RedoAction: FileAction
+{
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction
+	{
+		file.serverFileWrapper.redo()
+		return NoAction.tracedAction
+	}
+
+	override val type: FileActionType = FileActionType.REDO
+}
+
+/**
+ * `SaveAction` is a [FileAction] that saves an [AvailServerFile] to disk.
+ *
+ * @author Richard Arriaga &lt;rich@availlang.org&gt;
+ *
+ * @property failureHandler
+ *   A function that accepts a [ServerErrorCode] that describes the nature
+ *   of the failure and an optional [Throwable]. TODO refine error handling
+ *
+ * @constructor
+ * Construct a [SaveAction].
+ *
+ * @param failureHandler
+ *   A function that accepts a [ServerErrorCode] that describes the nature
+ *   of the failure and an optional [Throwable]. TODO refine error handling
+ */
+internal class SaveAction constructor(
+	private val failureHandler: (ServerErrorCode, Throwable?) -> Unit): FileAction
+{
+	override fun execute(file: AvailServerFile, timestamp: Long): TracedAction
+	{
+		file.save(failureHandler)
+		return NoAction.tracedAction
+	}
+
+	override val type: FileActionType = FileActionType.SAVE
+}
```

Comment:
As much as possible, I tried to write an API that was independent of what the client would do.  FileTypeActions may or may not be exposed to a client. This API is specific to the server. When we save we can perform a save as a synchronized file action. So this will likely be something only the server will call and it won't expose this API to the client.